udata/rdf.py

```diff
@@ -301,32 +301,35 @@ def contact_point_from_rdf(rdf, dataset):
             return contact_point or ContactPoint(name=name, email=email, owner=dataset.owner).save()
 
 
-def catalog_record_identifier_from_rdf(graph: Graph):
+def primary_topic_identifier_from_rdf(graph: Graph, resource: RdfResource):
     """
-    Extract the dct:identifier of a CatalogRecord from an RDF graph
+    Extract the dct:identifier from a primaryTopic of a RdfResource `resource` via an RDF `graph`.
+    The primary topic might be identified by a FOAF:isPrimaryTopicOf from the Resource, or by a FOAF:primaryTopic to the Resource.
+    In DCAT, the primary topic should be a unique CatalogRecord (if any), but nothing here prevents it to be something else.
     """
-    node_catalog_record = graph.value(predicate=RDF.type, object=DCAT.CatalogRecord)
-    if node_catalog_record:
-        catalog_record = graph.resource(node_catalog_record)
-        if catalog_record:
-            return rdf_value(catalog_record, DCT.identifier)
+    # look for "inner" primaryTopic linking to Dataset via isPrimaryTopicOf
+    is_primary_topic_of = graph.value(subject=resource.identifier, predicate=FOAF.isPrimaryTopicOf)
+    if is_primary_topic_of:
+        return graph.value(is_primary_topic_of, DCT.identifier)
+    # look for "outer" primaryTopic linking to Dataset via primaryTopic
+    primary_topic = graph.value(predicate=FOAF.primaryTopic, object=resource.identifier)
+    if primary_topic:
+        return graph.value(primary_topic, DCT.identifier)
 
 
 def remote_url_from_rdf(rdf: RdfResource, graph: Graph, remote_url_prefix: str | None = None):
     """
-    Compute from remote_url_prefix and CatalogRecord identifier if provided and found, respectively.
-    Or use DCAT.landingPage if found and uri validation succeeds.
-    Use RDF identifier as fallback if uri validation succeeds.
+    Compute from `remote_url_prefix` if provided and primaryTopic identifier if found.
+    Otherwise, use DCAT.landingPage if found and uri validation succeeds.
+    In this latter case, use RDF identifier as fallback if uri validation succeeds.
     """
-    if remote_url_prefix:
-        catalog_record_identifier = catalog_record_identifier_from_rdf(graph)
-        if catalog_record_identifier:
-            remote_url_prefix = (
-                f"{remote_url_prefix}/"
-                if not remote_url_prefix.endswith("/")
-                else remote_url_prefix
-            )
-            return f"{remote_url_prefix}{catalog_record_identifier}"
+    if remote_url_prefix and (
+        primary_topic_identifier := primary_topic_identifier_from_rdf(graph, rdf)
+    ):
+        remote_url_prefix = (
```

Comment:
Avoid the trailing slash gymnastics?
```
from urllib.parse import urljoin
return urljoin(remote_url_prefix, primary_topic_identifier)
```

Reply:
Nice idea, but it doesn't work with our canonical case : `http://catalogue.geo-ide.developpement-durable.gouv.fr/catalogue/srv/fre/catalog.search#/metadata/` + `fr-120066022-ldd-56fce164-04b2-41ae-be87-9f256f39dd44` (the `#` is ignored, as it should be on server side AFAIK).